erpnext/selling/doctype/sales_order/sales_order.py

```diff
@@ -241,14 +241,24 @@ def check_nextdoc_docstatus(self):
 		# Checks Sales Invoice
 		submit_rv = frappe.db.sql_list("""select t1.name
 			from `tabSales Invoice` t1,`tabSales Invoice Item` t2
-			where t1.name = t2.parent and t2.sales_order = %s and t1.docstatus < 2""",
+			where t1.name = t2.parent and t2.sales_order = %s and t1.docstatus = 1""",
 			self.name)
 
 		if submit_rv:
 			submit_rv = [get_link_to_form("Sales Invoice", si) for si in submit_rv]
 			frappe.throw(_("Sales Invoice {0} must be cancelled before cancelling this Sales Order")
 				.format(", ".join(submit_rv)))
 
+		draft_rv = frappe.db.sql_list("""select distinct t1.name
```

Comment:
Instead of running two queries, just get both of them in a single query and just also select doc status with them. Post query you can segregate the invoices and show the validation message accordingly

Reply:
Should I keep the draft_rv thing and get rid of everything else in the function? Because all the other conditions check for submitted docs, which means the errors won't ever be thrown, because the Cancel All dialog box would show up.